validphys2/src/validphys/utils.py

```diff
@@ -5,14 +5,52 @@
 @author: Zahari Kassabov
 """
 import contextlib
+import functools
 import pathlib
 import shutil
 import tempfile
+from typing import Any, Sequence, Mapping
 
 import numpy as np
 from validobj import ValidationError, parse_input
 
 
+# Since typing.Hashable doesn't check recursively you actually
+# have to try hashing it.
+def is_hashable(obj):
+    try:
+        hash(obj)
+        return True
+    except:
+        return False
+
+
+def immute(obj: Any):
+    # So that we don't infinitely recurse since frozenset and tuples
+    # are Sequences.
+    if is_hashable(obj):
+        return obj
+    elif isinstance(obj, Mapping):
+        return frozenset([(immute(k), immute(v)) for k, v in obj.items()])
+    elif isinstance(obj, Sequence):
+        return frozenset([immute(i) for i in obj])
+    else:
+        raise ValueError("Object is not hashable")
+
+
+def freeze_args(func):
+    """Transform mutable dictionary
+    Into immutable
+    Useful to be compatible with cache
+    """
+    @functools.wraps(func)
+    def wrapped(*args, **kwargs):
+        args = tuple([immute(arg) for arg in args])
+        kwargs = {k: immute(v) for k, v in kwargs.items()}
+        return func(*args, **kwargs)
```

Comment:
I feel this is rather dangerous as a general decorator, as it changes the meaning of the inputs in unexpected ways.

It may make more sense to do this closer to the runcard input. In which case the freezer becomes

```
def freeze(obj):
    if isinstance(obj, list):
        return tuple(freeze(ele) for ele in obj)
    if isinstance(obj, dict):
        return ((k, freeze(v)) for k, v in obj.items())
    if isinstance(obj, set):
        return frozenset(obj)
    return obj
```

Reply:
Right, but where in the code should the freezer be called?

Reply:
An input to a vp action is either:

  - Some input coming from the runcard, which is not hashable in general.
  - The ouput of a parsing or a production rule (like a `PDF` object) which really needs to be hashable already.

Hence you really only need to make sure to make the inputs hashable, which would not change the semantics too much. And the best place to do that is close to where you are parsing the input. It is something I had around for reportengine "1.0" which is why I had the above function lying around.